docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,484 @@
+
+        impl<T, U: CoreTryFrom<T>> SealedTryFrom<T> for U {}
+    }
+
+    pub trait Into<T: ?Sized>: sealed::SealedInto<T> {
+        fn into(&self) -> Box<T>;
+    }
+
+    impl<T, U: Clone + CoreInto<T>> Into<T> for U {
+        fn into(&self) -> Box<T> {
+            Box::new(self.clone().into())
+        }
+    }
+
+    pub trait TryFrom<T>: sealed::SealedTryFrom<T> {
+        type Error;
+
+        fn try_from(t: T) -> Result<Self, Self::Error>
+            where
+                Self: Sized;
+    }
+
+    impl<T, U: CoreTryFrom<T>> TryFrom<T> for U {
+        type Error = <Self as CoreTryFrom<T>>::Error;
+
+        fn try_from(t: T) -> Result<Self, Self::Error>
+            where
+                Self: Sized,
+        {
+            <Self as CoreTryFrom<T>>::try_from(t)
+        }
+    }
+}
+
+pub trait Protobuf<T: Message + Default>
+    where
+        Self: erased::TryFrom<T> + erased::Into<T>,
+        <Self as erased::TryFrom<T>>::Error: Display,
+{ /* ... */ }
+
+pub trait Header: Protobuf<Any> {}
+```
+
+#### `Clone` cannot be a supertrait and cannot be derived for types containing boxed trait objects
+
+The following can be made to work using the `dyn-clone` crate ->
+
+```rust
+#[derive(Clone)]
+pub struct Result {
+    pub client_id: ClientId,
+    pub client_state: Box<dyn ClientState>,
+    pub consensus_state: Box<dyn ConsensusState>,
+}
+```
+
+```rust
+pub trait ClientState: dyn_clone::DynClone { /* ... */ }
+
+dyn_clone::clone_trait_object!(ClientState);
+```
+
+`dyn-clone` also provides a `clone_box()` function that can be used to get a `Box<T>` from a `&T`.
+
+#### Deriving `serde::{Serialize, Deserialize}` for trait objects
+
+Replacing the `AnyHeader` with a `Box<dyn Header>` wouldn't work in the code below because `Deserialize`
+requires `Self: Sized` and `Serialize` uses generic types.
+
+```rust
+#[derive(Deserialize, Serialize)]
+pub struct UpdateClient {
+    pub common: Attributes,
+    pub header: Option<AnyHeader>,
+}
+```
+
+This can be solved using the `ibc_proto::google::protobuf::Any` type instead thereby deferring the actual
+encoding/decoding to the hosts themselves during handler execution.
+Thankfully, the core modules code doesn't use the `serde` derivations except in logs and errors. Host and light-client
+implementations can optionally choose to downcast to the concrete type and use it's `serde` derivations directly (if
+available).
+
+#### Light client traits cannot have constructors
+
+This restriction comes from the fact that trait methods of object safe traits cannot return `Self`.  
+However, we would need a constructor to be able to create a `ClientState` and `ConsensusState` in the `create_client`
+handler. This can be done using a `where Self: Sized` clause on the trait method.
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait ClientState {
+    fn decode(any: Any) -> Box<dyn ClientState> where Self: Sized;
+
+    /* ... */
+}
+```
+
+### Downcasting support
```

Comment:
Why would we ever need to downcast to a specific concrete type? Isn't the interface provided by, say, `Header` enough?

Reply:
I believe this will be required (for example) in cases where the light client wants to access its state stored on the host - the host is oblivious to the light client state's concrete type so it would return a trait object (such as `&dyn ConsensusState`) and the light client would need to downcast it to the concrete light client type (e.g. `TmConsensusState`) before it can use light client specific fields from it (such as `next_validators_hash`).